Common/src/Auth/Authentication/Authenticator.cs

```diff
@@ -150,20 +156,22 @@ protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
       return AuthenticateResult.Fail("Missing Certificate Headers");
     }
 
-    if (impersonationHeader_ != null
-        && Request.Headers.TryGetValue(impersonationHeader_, out var imps)
-        && !string.IsNullOrWhiteSpace(imps.First()))
+    var impersonationUsername = TryGetHeader(impersonationUsernameHeader_);
+    var impersonationId       = TryGetHeader(impersonationIdHeader_);
+
+    if (impersonationId != null || impersonationUsername != null)
     {
       if (identity.HasClaim(c => c.Type == Permissions.Impersonate.Claim.Type))
       {
-        identity = await GetImpersonationatedIdentity(identity,
-                                                      imps.First())
+        (identity, var message) = await GetImpersonatedIdentityAsync(identity!,
```

Comment:
do not overwrite a non nullable variable with a nullable object. You could create a new variable like `impersonated_identity`, and perform some checks on this one.

Reply:
They are both nullable, but I'll rewrite this part

Reply:
My point is: the function could return only a nullable string, and throw an exception with the message within in case of error.